File: sort/selection_sort/dart/selection_sort.dart
/// Selection Sort the input list
void selectionSort(List<int> items) {
  // Iterate through the array, knowing left-of-i is sorted and looking
  // for the lowest to-the-right element to place at the end of left
  for (int i = 0; i < items.length - 1; i++) {
    int lowestIndex = i;
    for (int j = i + 1; j < items.length; j++) {
      if (items[j] < items[lowestIndex]) {
        lowestIndex = j;
      }
    }

    // Swap the lowest item with this position (i)
    final int lowestNumber = items[lowestIndex];
    items[lowestIndex] = items[i];
    items[i] = lowestNumber;
  }
}
 // ----- Test -----
void main() {
  // 1. Arrange
  final List<int> testedList = [10, 2, 8, 5, 6, 1, 4, 7, 9, 3];
  final List<int> expectedOutput = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
   // 2. Act
  selectionSort(testedList);
  print(testedList);
   // 3. Assert
  // No simple list equality in dart yet so using 'fold' to recursively check
  // each sorted element compared to the expected output at each index.
  int i = 0;
  final bool isEqual = testedList.fold(
    true,
    (result, item) => result && item == expectedOutput[i++],
  );
  print(isEqual);
}

File: sort/selection_sort/swift/SelectionSort.swift
func selectionSort(unsortArr: [Int]) -> [Int] {
    var cloneArr = unsortArr

    for i in 0...cloneArr.count-1 {
        var minPos = i // min position
        for j in i...cloneArr.count-1 {
            if (cloneArr[j] < cloneArr[minPos]) {
                minPos = j
            }
        }

        let temp = cloneArr[i]
        cloneArr[i] = cloneArr[minPos]
        cloneArr[minPos] = temp
    }

    return cloneArr
}

var unsortArr1 = [3,1,6,2,4,7,5]
var sortedArr1 = selectionSort(unsortArr: unsortArr1)
print("test 1: input", unsortArr1)
print("test 1: output", sortedArr1)

print("---------------------")

var unsortArr2 = [9,5,8,2,3,9,9,10,4]
var sortedArr2 = selectionSort(unsortArr: unsortArr2)
print("test 2: input", unsortArr2)
print("test 2: output", sortedArr2)
